crates/workspace/src/workspace.rs

```diff
@@ -3122,21 +3123,46 @@ impl Workspace {
         }))
     }
 
+    pub fn join_all_panes(&mut self, cx: &mut ViewContext<Self>) {
+        let active_item = self.active_pane.read(cx).active_item().take();
+        for pane in self.panes.clone() {
+            self.move_all_items(cx, pane.clone(), self.active_pane.clone().clone());
+        }
+        if let Some(active_item) = active_item {
+            self.activate_item(&(*active_item), true, true, cx);
+        }
+        cx.notify();
+    }
+
     pub fn join_pane_into_next(&mut self, pane: View<Pane>, cx: &mut ViewContext<Self>) {
-        let next_pane = self
-            .find_pane_in_direction(SplitDirection::Right, cx)
-            .or_else(|| self.find_pane_in_direction(SplitDirection::Down, cx))
-            .or_else(|| self.find_pane_in_direction(SplitDirection::Left, cx))
-            .or_else(|| self.find_pane_in_direction(SplitDirection::Up, cx));
-        let Some(next_pane) = next_pane else {
+        let Some(next_pane) = self.find_next_pane(cx) else {
             return;
         };
+        self.move_all_items(cx, pane, next_pane);
+        cx.notify();
+    }
 
-        let item_ids: Vec<EntityId> = pane.read(cx).items().map(|item| item.item_id()).collect();
+    fn move_all_items(
+        &mut self,
+        cx: &mut ViewContext<'_, Workspace>,
+        from_pane: View<Pane>,
+        to_pane: View<Pane>,
+    ) {
+        let item_ids: Vec<EntityId> = from_pane
+            .read(cx)
+            .items()
+            .map(|item| item.item_id())
+            .collect();
         for item_id in item_ids {
-            self.move_item(pane.clone(), next_pane.clone(), item_id, 0, cx);
+            self.move_item(from_pane.clone(), to_pane.clone(), item_id, 0, cx);
```

Comment:
It's quite odd and sub optimal that first, we get all IDs and collect them, allocating memory

```rs
let item_ids: Vec<EntityId> = from_pane
    .read(cx)
    .items()
    .map(|item| item.item_id())
    .collect();
```


only to re-iterate every pane items again, to get its position number

```rs
let Some((item_ix, item_handle)) = source
    .read(cx)
    .items()
    .enumerate()
    .find(|(_, item_handle)| item_handle.item_id() == item_id_to_move)
else {
    // Tab was closed during drag
    return;
};
```

This is plain `O(n^2)` and we can do much better.

To continue, `self` is not needed in this method at all, and `cx` is only used to fetch those IDs — removing both will allow to get rid of `.collect()` call above and another allocation.